### PR TITLE
Advance permanent slots for cascading nodes while in failsafe

### PR DESCRIPTION
**Pull request by CyberDem0n**
Lets consider a following replication setup:
```
primary->standby1->standby2(replicatefrom: standby1)
```

In this case the `primary` will not create a physical replication slot for standby2, because it is streaming from the `standby1`.

Things will look differently if we have the following dynamic configuration:
```yaml
slots:
    primary:
        type: physical
    standby1:
        type: physical
    standby2:
        type: physical
```

In this case `primary` will also have `standby2` physical replication slot, which periodically must be advanced. So far it was working by taking value of `xlog_location` from the `/members/standby2` key in DCS.

But, when DCS is down and failsafe mode is activate, the `standby2` physical slot on the `primary` will not not be moved, because there was not way to get the latest value of `xlog_location`.

This PR is addressing the problem by making replica nodes to return their `xlog_location` as `lsn` header in the response on `POST /failsafe` REST API request. The current primary will use these values to advance replication slots for nodes with `replicatefrom` tag.

### CODE REVIEW

**Comment by barthisrael** [comment]
The implementation seems to work fine.

I performed a quick test by setting up cascading replication with the configuration files available in the repository (`postgresql0.yml`, `postgresql1.yml` and `postgresql2.yml` with `replicatefrom` uncommendted).

* Dynamic config:

```console
$ patronictl -c postgres0.yml show-config
failsafe_mode: true
loop_wait: 10
maximum_lag_on_failover: 1048576
postgresql:
  parameters: null
  pg_hba:
  - host replication replicator 127.0.0.1/32 md5
  - host all all 0.0.0.0/0 md5
  use_pg_rewind: true
retry_timeout: 10
slots:
  postgresql0:
    type: physical
  postgresql1:
    type: physical
  postgresql2:
    type: physical
ttl: 30
```

* Cluster topology

```console
$ patronictl -c postgres0.yml topology
+ Cluster: batman (7392576939261466290) -----+-----------+----+-----------+------------------------------+
| Member          | Host           | Role    | State     | TL | Lag in MB | Tags                         |
+-----------------+----------------+---------+-----------+----+-----------+------------------------------+
| postgresql0     | 127.0.0.1:5432 | Leader  | running   |  9 |           |                              |
| + postgresql1   | 127.0.0.1:5433 | Replica | streaming |  9 |         0 |                              |
|   + postgresql2 | 127.0.0.1:5434 | Replica | streaming |  9 |         0 | {replicatefrom: postgresql1} |
+-----------------+----------------+---------+-----------+----+-----------+------------------------------+
```

* Slots status:

```console
$ PGPASSWORD=patroni psql -h localhost -U postgres -c "SELECT slot_name, active, restart_lsn FROM  pg_replication_slots ORDER BY slot_name"
  slot_name  | active | restart_lsn
-------------+--------+-------------
 postgresql1 | t      | 0/CAD09A0
 postgresql2 | f      | 0/CAD09A0
(2 rows)
```

* Stop etcd

```console
$ sudo systemctl stop etcd
```

* Check we are in failsafe mode

```console
$ curl -s http://localhost:8008/patroni | jq '.failsafe_mode_is_active'
true
```

* Put some load

```console
$ PGPASSWORD=patroni psql -h localhost -U postgres -c "CREATE TABLE test AS SELECT generate_series(1, 1000000); DROP TABLE test;"
SELECT 1000000
DROP TABLE
```

* Check both slots are advancing

```console
$ PGPASSWORD=patroni psql -h localhost -U postgres -c "SELECT slot_name, active, restart_lsn FROM  pg_replication_slots ORDER BY slot_name"
  slot_name  | active | restart_lsn
-------------+--------+-------------
 postgresql1 | t      | 0/1081CBF8
 postgresql2 | f      | 0/1081CBF8
(2 rows)
```

File: patroni/ha.py [hunk 1]
```diff
@@ -88,14 +88,53 @@ def failover_limitation(self) -> Optional[str]:
         return None
 
 
+class _FailsafeResponse(NamedTuple):
+    """Response on POST /failsafe API request.
```
Comment (barthisrael) [inline patroni/ha.py:92]:
```suggestion
    """Response on POST ``/failsafe`` API request.
```

File: patroni/ha.py [hunk 2]
```diff
@@ -88,14 +88,53 @@ def failover_limitation(self) -> Optional[str]:
...
+    lsn: Optional[int]
+
+
 class Failsafe(object):
+    """Object that represents failsafe state of the cluster."""
 
     def __init__(self, dcs: AbstractDCS) -> None:
+        """Initialize the :class:`Failsafe` object.
+
+        :param dcs: current DCS object, is used only to get current value of `ttl`.
```
Comment (barthisrael) [inline patroni/ha.py:112]:
```suggestion
        :param dcs: current DCS object, is used only to get current value of ``ttl``.
```

File: patroni/ha.py [hunk 3]
```diff
@@ -88,14 +88,53 @@ def failover_limitation(self) -> Optional[str]:
...
+        """
         self._lock = RLock()
         self._dcs = dcs
         self._reset_state()
 
+    def update_slots(self, slots: Dict[str, int]) -> None:
+        """Assign value to `_slots`.
```
Comment (barthisrael) [inline patroni/ha.py:119]:
```suggestion
        """Assign value to :attr:`_slots`.
```

File: patroni/ha.py [hunk 4]
```diff
@@ -88,14 +88,53 @@ def failover_limitation(self) -> Optional[str]:
...
+
+        :param slots: a :class:`dict` object with member names as keys and received/replayed LSNs as values.
+        """
+        with self._lock:
+            self._slots = slots
+
     def update(self, data: Dict[str, Any]) -> None:
+        """Update the :class:`Failsafe` object state.
+
+        The last update time is stored and object will be invalidated after `ttl` seconds.
```
Comment (barthisrael) [inline patroni/ha.py:131]:
```suggestion
        The last update time is stored and object will be invalidated after ``ttl`` seconds.
```

File: patroni/ha.py [hunk 5]
```diff
@@ -104,44 +143,74 @@ def update(self, data: Dict[str, Any]) -> None:
...
                 return Leader('', '', RemoteMember(self._name, {'api_url': self._api_url,
                                                                 'conn_url': self._conn_url,
                                                                 'slots': self._slots}))
 
     def update_cluster(self, cluster: Cluster) -> Cluster:
+        """Update and return provided :class:`Cluster` object with fresh values.
+
+        .. note::
+            This method is called when failsafe mode is active and is used to update cluster state
+            with fresh values of replication `slots` status and `xlog_location` on member nodes.
```
Comment (barthisrael) [inline patroni/ha.py:167]:
```suggestion
            with fresh values of replication ``slots`` status and ``xlog_location`` on member nodes.
```

File: patroni/ha.py [hunk 6]
```diff
@@ -104,44 +143,74 @@ def update(self, data: Dict[str, Any]) -> None:
...
-           On primary the self._last_update is set from the
-           set_is_active() method and always returns the correct value.
+        .. note:
+            This method is called from the REST API to report whether the failsafe mode was activated.
 
-           On replicas the self._last_update is set at the moment when
-           the primary performs POST /failsafe REST API calls.
-           The side-effect - it is possible that replicas will show
-           failsafe_is_active values different from the primary."""
+            On primary the self._last_update is updated from the :func:`set_is_active` method and always
```
Comment (barthisrael) [inline patroni/ha.py:190]:
```suggestion
            On primary the :attr:`_last_update` is updated from the :func:`set_is_active` method and always
```

File: patroni/ha.py [hunk 7]
```diff
@@ -104,44 +143,74 @@ def update(self, data: Dict[str, Any]) -> None:
...
+            returns the correct value.
+
+            On replicas the self._last_update is updated at the moment when the primary performs
```
Comment (barthisrael) [inline patroni/ha.py:193]:
```suggestion
            On replicas the :attr:`_last_update` is updated at the moment when the primary performs
```

File: patroni/ha.py [hunk 8]
```diff
@@ -104,44 +143,74 @@ def update(self, data: Dict[str, Any]) -> None:
...
+            `POST /failsafe` REST API calls.
```
Comment (barthisrael) [inline patroni/ha.py:194]:
```suggestion
            ``POST /failsafe`` REST API calls.
```

File: patroni/ha.py [hunk 9]
```diff
@@ -104,44 +143,74 @@ def update(self, data: Dict[str, Any]) -> None:
...
+            values different from the primary.
+
+        :returns: ``True`` if failsafe mode is active, ``False`` otherwise.
+        """
 
         with self._lock:
             return self._last_update + self._dcs.ttl > time.time()
 
     def set_is_active(self, value: float) -> None:
+        """Update self._last_update value.
```
Comment (barthisrael) [inline patroni/ha.py:206]:
```suggestion
        """Update :attr:`_last_update` value.
```

File: patroni/ha.py [hunk 10]
```diff
@@ -104,44 +143,74 @@ def update(self, data: Dict[str, Any]) -> None:
...
+
+        .. note::
+            This method is only called on the primary.
+            Effectively it sets expiration time of failsafe mode. If the provided value is 0, it disables failsafe mode.
```
Comment (barthisrael) [inline patroni/ha.py:210]:
```suggestion
            Effectively it sets expiration time of failsafe mode. If the provided value is ``0``, it disables failsafe mode.
```

File: patroni/ha.py [hunk 11]
```diff
@@ -921,23 +996,38 @@ def fetch_nodes_statuses(self, members: List[Member]) -> List[_MemberStatus]:
...
             return 'Running as a leader'
         self._failsafe.update(data)
+        return self._last_wal_lsn
 
     def failsafe_is_active(self) -> bool:
         return self._failsafe.is_active()
 
-    def call_failsafe_member(self, data: Dict[str, Any], member: Member) -> bool:
+    def call_failsafe_member(self, data: Dict[str, Any], member: Member) -> _FailsafeResponse:
+        """Call `POST /failsafe` REST API request on provided member.
```
Comment (barthisrael) [inline patroni/ha.py:1015]:
```suggestion
        """Call ``POST /failsafe`` REST API request on provided member.
```

**Review by barthisrael — COMMENTED** [review]
Code changes look good.
Posted some suggestions in regards to sphinx recommendations for docstrings.

**Comment by CyberDem0n** [comment]
Thank you @barthisrael for a review!

**Comment by barthisrael** [comment]
👍

**Review by barthisrael — APPROVED** [review]

**Review by hughcapet — APPROVED** [review]

**Pull request merged**